### PR TITLE
feat(web): add device test page

### PR DESCRIPTION
**Pull request by Cheerego7**
feat(flat-components):  add CameraTest, SpeakerTes, MicrophoneTest, DeviceTestPanel component to flat-components

### CODE REVIEW

File: packages/flat-components/src/components/DeviceTestPage/CameraTest/index.tsx [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+            </div>
+            <div className="camera-test-wrapper">
+                <div className="camera-box" ref={cameraVideoStreamRef} />
+                <div
+                    className={classNames("camera-no-accredit-box", {
+                        visible: !isCameraAccessible,
+                    })}
+                >
+                    <img src={cameraDisabledSVG} />
+                    <span>请开启浏览器的摄像头使用权限</span>
```
Comment (BlackHole1) [inline packages/flat-components/src/components/DeviceTestPage/CameraTest/index.tsx]: i18n

File: packages/flat-components/src/components/DeviceTestPage/CameraTest/index.tsx [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+export const CameraTest: React.FC<CameraTestProps> = ({
+    cameraDevices,
+    isCameraAccessible,
+    currentCameraDeviceID,
+    cameraVideoStreamRef,
+    setCameraDevice,
+}) => {
+    return (
+        <div className="camera-test-container">
+            <div className="camera-test-text">摄像头</div>
```
Comment (BlackHole1) [inline packages/flat-components/src/components/DeviceTestPage/CameraTest/index.tsx]: i18n

File: packages/flat-components/src/components/DeviceTestPage/CameraTest/CameraTest.stories.tsx [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+
+export const PlayableExample: Story<CameraTestProps> = args => (
+    <div className="vh-100 mw5-ns">
+        <CameraTest {...args} />
+    </div>
+);
+PlayableExample.args = {
+    cameraDevices: [
+        { label: "FaceTime HD Camera", deviceId: "1", groupId: "1", kind: "audioinput" },
+        { label: "other camera", deviceId: "2", groupId: "2", kind: "audioinput" },
```
Comment (BlackHole1) [inline packages/flat-components/src/components/DeviceTestPage/CameraTest/CameraTest.stories.tsx]: wrap line

File: packages/flat-components/src/components/DeviceTestPage/DeviceTestPanel.stories.tsx [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+export const PlayableExample: Story<DeviceTestPanelProps> = args => (
+    <div className="vh-100 w-70">
+        <DeviceTestPanel {...args} />
+    </div>
+);
+PlayableExample.args = {
+    cameraDevices: [
+        { label: "FaceTime HD Camera", deviceId: "1", groupId: "1", kind: "audioinput" },
+        { label: "other camera", deviceId: "2", groupId: "2", kind: "audioinput" },
```
Comment (BlackHole1) [inline packages/flat-components/src/components/DeviceTestPage/DeviceTestPanel.stories.tsx]: wrap line

File: packages/flat-components/src/components/DeviceTestPage/DeviceTestPanel.stories.tsx [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+    ],
+    speakerDevices: [
+        { label: "default(MacBook Pro speaker)", deviceId: "1", groupId: "1", kind: "audioinput" },
+        { label: "other speaker", deviceId: "2", groupId: "2", kind: "audioinput" },
```
Comment (BlackHole1) [inline packages/flat-components/src/components/DeviceTestPage/DeviceTestPanel.stories.tsx]: wrap line

File: packages/flat-components/src/components/DeviceTestPage/DeviceTestPanel.stories.tsx [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+    ],
+    microphoneDevices: [
+        {
+            label: "default(MacBook Pro Microphone)",
+            deviceId: "1",
+            groupId: "1",
+            kind: "audioinput",
+        },
+        { label: "other Microphone", deviceId: "2", groupId: "2", kind: "audioinput" },
```
Comment (BlackHole1) [inline packages/flat-components/src/components/DeviceTestPage/DeviceTestPanel.stories.tsx]: wrap line

File: packages/flat-components/src/components/DeviceTestPage/DeviceTestSelect/DeviceTestSelect.stories.tsx [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+
+export const PlayableExample: Story<DeviceTestSelectProps> = args => (
+    <div className="vh-100 mw5-ns">
+        <DeviceTestSelect {...args} />
+    </div>
+);
+PlayableExample.args = {
+    devices: [
+        { label: "FaceTime HD Camera", deviceId: "1", groupId: "1", kind: "audioinput" },
+        { label: "other camera", deviceId: "2", groupId: "2", kind: "audioinput" },
```
Comment (BlackHole1) [inline packages/flat-components/src/components/DeviceTestPage/DeviceTestSelect/DeviceTestSelect.stories.tsx]: wrap line

File: packages/flat-components/src/components/DeviceTestPage/DeviceTestSelect/index.tsx [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+                                <Select.Option value={deviceId} key={deviceId}>
+                                    {label}
+                                </Select.Option>
+                            );
+                        })}
+                    </Select>
+                </div>
+            ) : (
+                <div className="device-test-select-disabled-btn">
+                    <Button icon={<img src={disabledSVG} />}>未开启权限</Button>
```
Comment (BlackHole1) [inline packages/flat-components/src/components/DeviceTestPage/DeviceTestSelect/index.tsx]: i18n

File: packages/flat-components/src/components/DeviceTestPage/MicrophoneTest/MicroponeTest.stories.tsx [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+);
+PlayableExample.args = {
+    microphoneDevices: [
+        {
+            label: "default(MacBook Pro Microphone)",
+            deviceId: "1",
+            groupId: "1",
+            kind: "audioinput",
+        },
+        { label: "other Microphone", deviceId: "2", groupId: "2", kind: "audioinput" },
```
Comment (BlackHole1) [inline packages/flat-components/src/components/DeviceTestPage/MicrophoneTest/MicroponeTest.stories.tsx]: wrap line

File: packages/flat-components/src/components/DeviceTestPage/MicrophoneTest/index.tsx [hunk 10]
```diff
@@ -0,0 +1,42 @@
...
+export const MicrophoneTest: React.FC<MicrophoneTestProps> = ({
+    microphoneDevices,
+    microphoneVolume,
+    isMicrophoneAccessible,
+    currentMicrophoneDeviceID,
+    setMicrophoneDevice,
+}) => {
+    return (
+        <div className="microphone-test-container">
+            <div className="microphone-test-text">麦克风</div>
```
Comment (BlackHole1) [inline packages/flat-components/src/components/DeviceTestPage/MicrophoneTest/index.tsx]: i18n

**Review by BlackHole1 — CHANGES_REQUESTED** [review]
wrap line and i18n

File: web/flat-web/src/api-middleware/rtc/device-test.ts [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+export class DeviceTest {
+    public devices: MediaDeviceInfo[] = [];
+    public onDevicesChanged?: (devices: MediaDeviceInfo[]) => void;
+
+    public cameraVideoTrack?: ICameraVideoTrack;
+    public microphoneAudioTrack?: IMicrophoneAudioTrack;
+
+    public volume = 0;
+    public onVolumeChanged?: (volume: number) => void;
+    private volumeTimer = 0;
```
Comment (BlackHole1) [inline web/flat-web/src/api-middleware/rtc/device-test.ts]:
```suggestion
    private volumeTimer = NaN;
```

File: web/flat-web/src/pages/DevicesTestPage/index.tsx [hunk 12]
```diff
@@ -0,0 +1,179 @@
...
+import { joinRoomHandler } from "../utils/join-room-handler";
+import { GlobalStoreContext } from "../../components/StoreProvider";
+
+export const DevicesTestPage = observer(function DeviceTestPage() {
+    const globalStore = useContext(GlobalStoreContext);
+    const { roomUUID } = useParams<RouteParams<RouteNameType.JoinPage>>();
+    const pushHistory = usePushHistory();
+
+    const audioResourceSrc =
+        "https://flat-storage.oss-accelerate.aliyuncs.com/cloud-storage/2021-11/25/df1ba448-643c-4b31-8b92-f9439d5d9389/df1ba448-643c-4b31-8b92-f9439d5d9389.mp3";
```
Comment (BlackHole1) [inline web/flat-web/src/pages/DevicesTestPage/index.tsx]: use local static. e.g. https://github.com/netless-io/flat/blob/main/desktop/renderer-app/src/assets/media/Goldberg%20Variations%2C%20BWV%20988%20-%2005%20-%20Variatio%204%20a%201%20Clav.mp3

We need to be aware of copyright issues.

File: web/flat-web/src/route-config.ts [hunk 13]
```diff
@@ -86,6 +87,9 @@ export const routeConfig = {
     [RouteNameType.ResourcePreviewPage]: {
         path: "/preview/:fileURL/:taskToken?/:taskUUID?/:region?/",
     },
+    [RouteNameType.DevicesTestPage]: {
+        path: "/test/:roomUUID/",
```
Comment (BlackHole1) [inline web/flat-web/src/route-config.ts]: devices-check

**Review by BlackHole1 — CHANGES_REQUESTED** [review]

**Comment by LitoMore** [comment]
SVGs need to be optimized.

You can try with [SVGO](https://github.com/svg/svgo#other-ways-to-use-svgo).

**Comment by BlackHole1** [comment]
> SVGs need to be optimized.
> 
> You can try with [SVGO](https://github.com/svg/svgo#other-ways-to-use-svgo).

This should be done in another PR as all the svg's in this project need to be optimised.

File: web/flat-web/src/pages/UserSettingPage/GeneralSettingPage/index.tsx [hunk 14]
```diff
@@ -36,6 +39,17 @@ export const GeneralSettingPage = (): React.ReactElement => {
...
                 </div>
+                <div className="general-setting-device-test-box">
+                    <div className="general-setting-checkbox-title">设备检测选项</div>
+                    <Checkbox
+                        defaultChecked={!globalStore.isCloseDeviceTest}
+                        onClick={() => {
+                            globalStore.switchDeviceTest();
+                        }}
+                    >
+                        进入房间时开启
```
Comment (BlackHole1) [inline web/flat-web/src/pages/UserSettingPage/GeneralSettingPage/index.tsx]: i18n

File: web/flat-web/src/pages/UserSettingPage/GeneralSettingPage/index.tsx [hunk 15]
```diff
@@ -36,6 +39,17 @@ export const GeneralSettingPage = (): React.ReactElement => {
                         <Radio value={SelectLanguage.English}>English</Radio>
                     </Radio.Group>
                 </div>
+                <div className="general-setting-device-test-box">
+                    <div className="general-setting-checkbox-title">设备检测选项</div>
```
Comment (BlackHole1) [inline web/flat-web/src/pages/UserSettingPage/GeneralSettingPage/index.tsx]: i18n

File: packages/flat-components/src/components/DeviceTestPage/index.tsx [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+                    />
+                </div>
+            </div>
+            <div className="device-test-panel-tips-box">
+                <div className="device-test-panel-tips-radio">
+                    <Checkbox onClick={closeDeviceTest}>不再提示</Checkbox>
+                </div>
+                <div className="device-test-panel-join-btn">
+                    <Button type="primary" onClick={joinRoom}>
+                        进入房间
```
Comment (BlackHole1) [inline packages/flat-components/src/components/DeviceTestPage/index.tsx]: i18n

File: packages/flat-components/src/components/DeviceTestPage/index.tsx [hunk 17]
```diff
@@ -0,0 +1,96 @@
...
+                        isMicrophoneAccessible={isMicrophoneAccessible}
+                        microphoneVolume={microphoneVolume}
+                        currentMicrophoneDeviceID={currentMicrophoneDeviceID}
+                        setMicrophoneDevice={setMicrophoneDevice}
+                    />
+                </div>
+            </div>
+            <div className="device-test-panel-tips-box">
+                <div className="device-test-panel-tips-radio">
+                    <Checkbox onClick={closeDeviceTest}>不再提示</Checkbox>
```
Comment (BlackHole1) [inline packages/flat-components/src/components/DeviceTestPage/index.tsx]: i18n

File: web/flat-web/src/pages/DevicesTestPage/index.tsx [hunk 18]
```diff
@@ -0,0 +1,175 @@
...
+
+        if (refreshDevicesTimer.current) {
+            return;
+        }
+        refreshDevices();
+    }, [isCameraAccessible, deviceTest, isMicrophoneAccessible, refreshDevices]);
+
+    const closeDeviceTest = (): void => {
+        globalStore.switchDeviceTest();
+        console.log("is open device test", globalStore.isCloseDeviceTest);
```
Comment (BlackHole1) [inline web/flat-web/src/pages/DevicesTestPage/index.tsx]: `closeDeviceTest` / `switchDeviceTest` /  `is open device`?

Variable names are problematic

File: packages/flat-components/src/components/DeviceTestPage/index.tsx [hunk 19]
```diff
@@ -0,0 +1,96 @@
...
+    cameraVideoStreamRef,
+    setCameraDevice,
+    setMicrophoneDevice,
+    setSpeakerDevice,
+    closeDeviceTest,
+    joinRoom,
+}) => {
+    return (
+        <div className="device-test-panel-container">
+            <div className="device-test-panel-title-box">设备检测</div>
```
Comment (BlackHole1) [inline packages/flat-components/src/components/DeviceTestPage/index.tsx]: i18n

File: packages/flat-components/src/components/DeviceTestPage/SpeakerTest/index.tsx [hunk 20]
```diff
@@ -0,0 +1,61 @@
...
+    isSpeakerAccessible,
+    audioResourceSrc,
+    setSpeakerDevice,
+}) => {
+    const { t } = useTranslation();
+    const [isPlaying, setIsPlaying] = useState(false);
+    const audioRef = useRef<HTMLAudioElement>(null);
+
+    const togglePlay = (): void => {
+        !isPlaying ? audioRef.current?.play() : audioRef.current?.pause();
```
Comment (BlackHole1) [inline packages/flat-components/src/components/DeviceTestPage/SpeakerTest/index.tsx]:
```suggestion
        isPlaying ? audioRef.current?.pause() : audioRef.current?.play();
```

**Review by BlackHole1 — CHANGES_REQUESTED** [review]

**Comment by Cheerego7** [comment]
> SVGs need to be optimized.
> 
> You can try with [SVGO](https://github.com/svg/svgo#other-ways-to-use-svgo).

Thank you for sharing your advice with me, we can create issue of the SVG optimize

File: web/flat-web/src/pages/DevicesTestPage/index.tsx [hunk 21]
```diff
@@ -0,0 +1,174 @@
...
+            return;
+        }
+
+        if (refreshDevicesTimer.current) {
+            return;
+        }
+        refreshDevices();
+    }, [isCameraAccessible, deviceTest, isMicrophoneAccessible, refreshDevices]);
+
+    const turnOffDeviceTest = (): void => {
```
Comment (BlackHole1) [inline web/flat-web/src/pages/DevicesTestPage/index.tsx]: `turnOffDeviceTest` and `toggleOnAndOffDeviceTest`  Semantic conflict

File: web/flat-web/src/stores/GlobalStore.ts [hunk 22]
```diff
@@ -85,6 +86,10 @@ export class GlobalStore {
         this.isShowRecordHintTips = false;
     };
 
+    public toggleOnAndOffDeviceTest = (): void => {
+        this.isTurnOffDeviceTest = !this.isTurnOffDeviceTest;
+    };
```
Comment (BlackHole1) [inline web/flat-web/src/stores/GlobalStore.ts]:
```suggestion
    public toggleOnAndOffDeviceTest = (switch?: boolean): void => {
        this.isTurnOffDeviceTest = switch === undefined ? !this.isTurnOffDeviceTest : switch;
    };
```

**Review by BlackHole1 — CHANGES_REQUESTED** [review]

**Review by BlackHole1 — APPROVED** [review]

**Review by BlackHole1 — APPROVED** [review]

**Pull request merged**